src/DataStructures/VectorImpl.hpp

```diff
@@ -33,6 +33,46 @@
 #include "Utilities/StdArrayHelpers.hpp"
 #include "Utilities/TypeTraits/IsComplexOfFundamental.hpp"
 
+class ComplexDataVector;
+class ComplexModalVector;
+class DataVector;
+class ModalVector;
+
+namespace VectorImpl_detail {
+/// \brief Whether or not a given vector type is assignable to another
+///
+/// \details
+/// This is used to define which types can be assigned to one another. For
+/// example, you can assign a `ComplexDataVector` to a `DataVector`, but not
+/// vice versa.
+///
+/// To enable assignments between more types, modify a current template
+/// specialization or add a new one.
+///
+/// \tparam LhsDataType the type being assigned
+/// \tparam RhsDataType the type to convert to `LhsDataType`
```

Comment:
Can all this be checked with expression SFINAE? Like:
```cpp
template <typename T, typename S, std::void_t<>> is_assignable : std::false_type {};

template <typename T, typename S, std::void_t<decltype(std::declval<T>() = std::declval<S>())>> is_assignable :  std::true_type {};
```
and for that matter, https://en.cppreference.com/w/cpp/types/is_assignable might work even. The downside of `std::is_assignable` is that you need full type definitions at places where it's checked, but that should be fine. I'm not sure if it silently fails if the full type definition isn't available, but I assume so and is probably the correct behavior.

Reply:
I will look into this and use it if it works!

Reply:
I tried these but the issue I ran into is the opposite of what I would have expected - that I think more assignments would be permitted than we want. Apparently, you can assign a `DataVector` to a `ComplexDataVector`, but I imagine that we don't want that because when would we want that? And I figure in most cases if someone does that, it's an accident and the mistake would be hidden.

Because of this, I did not end up using either of the suggestions here in my fixups.